### ISSUE TITLE
[Meta] Wasm-based experimental plugin support

### ISSUE THREAD

**Issue by kwonoj** [swc-project/swc#3167]
### Describe the feature

Originating from https://github.com/swc-project/swc/issues/2635#issue-1043151813 . 

It looks like we are able to push forward wasm based plugin, this issue is filed for tracking its subtasks to officially release features. Still it'll be exposed under `experimental` flag to iterate implementaion with possibly breaking changes.

- [x] runtime replacement (https://github.com/swc-project/swc/pull/3196 / https://github.com/wasmerio/wasmer/issues/2324) : hope to use wasmer.
- [x] deprecate `abi_stable` (https://github.com/swc-project/swc/pull/3179)
- [x] deprecate `plugin_ast`
- [x] `swc_common/plugin` to utilize https://wasmerio.github.io/wasmer/crates/doc/wasmer/macro.imports.html
- [x] support loading / bytecode caching wasm from `swc_plugin_runner` (https://github.com/swc-project/swc/pull/3179)
- [x] configuration support for bytecode cache location (https://github.com/swc-project/swc/pull/3179)
- [ ] compiled bytecode cache support
- [ ] swc_cli supports for plugin project scafolding (we'll consolidate https://github.com/swc-project/swc-dev/blob/main/src/main.rs into `swc`'s subcommand)
- [x] writing short adr (https://github.com/swc-project/swc/pull/3249)
- [ ] Cache expiration strategy (Probably will not be implemented in experimental phase, need bit more thought. May spin off new issue later).
- [x] Optimize wasm loading strategy: current impl (https://github.com/swc-project/swc/blob/fc4c6708f24cda39640fbbfe56123f2f6eeb2474/crates/swc/src/plugin.rs#L19) is incorrect as it is being called per each file transform operation. We should read plugin binary only once per whole transform. (https://github.com/swc-project/swc/pull/3186)
- [x] println / dbg! or tracing support (https://github.com/swc-project/swc/pull/3271)
- [x] error handling (https://github.com/swc-project/swc/pull/3300)

**Comment by kwonoj** [comment]
Subtasks:

- https://github.com/swc-project/swc/issues/3311
- https://github.com/swc-project/swc/issues/3309
- https://github.com/swc-project/swc/issues/3308
- https://github.com/swc-project/swc/issues/3307

Now plugin implementation reached `at least working` stage, further tasks will be filed separately instead of using meta issue. This issue will be closed soon once a few remaining tasks here make some progress.

**Comment by kwonoj** [comment]
Closing as initial completion.

Please use https://github.com/swc-project/swc/discussions/3540 for the suggestions / discussions, and create a new issue if there's a bug need to be fixed.
